src/Microsoft.OData.Core/UriParser/Binders/SelectPathSegmentTokenBinder.cs

```diff
@@ -178,18 +178,24 @@ internal static bool TryBindAsOperation(PathSegmentToken pathToken, IEdmModel mo
             }
 
             // Only filter if there is more than one and its needed.
-            if (possibleFunctions.Count() > 1)
+            if(possibleFunctions.HasAny())
             {
-                possibleFunctions = possibleFunctions.FilterBoundOperationsWithSameTypeHierarchyToTypeClosestToBindingType(entityType);
-            }
+                int functionsCount = possibleFunctions.Count();
 
-            // If more than one overload matches, try to select based on optional parameters
-            if (possibleFunctions.Count() > 1 && parameterNames.Count > 0)
-            {
-                possibleFunctions = possibleFunctions.FilterOverloadsBasedOnParameterCount(parameterNames.Count);
-            }
+                if(functionsCount > 1)
+                {
+                    possibleFunctions = possibleFunctions.FilterBoundOperationsWithSameTypeHierarchyToTypeClosestToBindingType(entityType);
+
+                    int parameterNamesCount = parameterNames.Count;
 
-            if (!possibleFunctions.HasAny())
+                    // If more than one overload matches, try to select based on optional parameters
+                    if (parameterNamesCount > 0)

```

Comment:
I think we may have inadvertently changed the logic here; previously, this was only called if there was more than one possible function *after* calling FilterBoundOperationsWIthSameTypeHierarchyToTypeCLosestToBindingType() and parameterNames.Count > 0.  Now we call this even if there is only one match after calling FilterBoundOperationsWithSameTypeHierarchyToTypeClosestToBindingType().  To preserve the old logic, we have to confirm that possibleFunctions is still >1 after callingFilterBoundOperationsWIthSameTypeHierarchyToTypeClosestToBindingType().

Also, previous logic checked to see if there were any functions *after* filtering based on parameter count, where-as the new logic assumes there will be at least one function with the exact parameter count.

But, we should still only check this if the initial count > 1, and we should check parameterNames.Count before possibleFunctions.Count().  And, we probably only have to do the last Any() check if we tried to filter by parameter count.

The more I look at this, the more I think it needs to be entirely rewritten.  In the meantime, I think the following code is a slight improvement and preserves the existing logic:

            // Only filter if there is more than one and its needed.
            if (possibleFunctions.Count() > 1)
            {
                possibleFunctions = possibleFunctions.FilterBoundOperationsWithSameTypeHierarchyToTypeClosestToBindingType(entityType);

                // If there is still more than one overload that matches, try to select based on optional parameters
                if (parameterNames.Count > 0 && possibleFunctions.Count() > 1 )
                {
                    possibleFunctions = possibleFunctions.FilterOverloadsBasedOnParameterCount(parameterNames.Count);
                    if (!possibleFunctions.HasAny())
                    {
                        segment = null;
                        return false;
                    }
                }
            }
            else
            {
                segment = null;
                return false;
            }